fortuna/src/keeper.rs

```diff
@@ -0,0 +1,388 @@
+use {
+    crate::{
+        api::{
+            self,
+            BlockchainState,
+        },
+        chain::{
+            ethereum::SignablePythContract,
+            reader::{
+                BlockNumber,
+                RequestedWithCallbackEvent,
+            },
+        },
+        config::EthereumConfig,
+    },
+    anyhow::Result,
+    ethers::{
+        providers::{
+            Middleware,
+            Provider,
+            Ws,
+        },
+        types::U256,
+    },
+    futures::StreamExt,
+    std::sync::Arc,
+    tokio::{
+        spawn,
+        sync::mpsc,
+        time::{
+            self,
+            Duration,
+        },
+    },
+    tracing,
+};
+
+
+/// How much to wait before retrying in case of an RPC error
+const RETRY_INTERVAL: Duration = Duration::from_secs(5);
+/// How many blocks to look back for events that might be missed when starting the keeper
+const BACKLOG_RANGE: u64 = 1000;
+/// How many blocks to fetch events for in a single rpc call
+const BLOCK_BATCH_SIZE: u64 = 100;
+/// How much to wait before polling the next latest block
+const POLL_INTERVAL: Duration = Duration::from_secs(5);
+
+
+/// Get the latest safe block number for the chain. Retry internally if there is an error.
+async fn get_latest_safe_block(chain_config: &BlockchainState) -> BlockNumber {
+    loop {
+        match chain_config
+            .contract
+            .get_block_number(chain_config.confirmed_block_status)
+            .await
+        {
+            Ok(latest_confirmed_block) => {
+                return latest_confirmed_block - chain_config.reveal_delay_blocks
+            }
+            Err(e) => {
+                tracing::error!("Error while getting block number. error: {:?}", e);
+                time::sleep(RETRY_INTERVAL).await;
+            }
+        }
+    }
+}
+
+
+pub async fn run_keeper_threads(
```

Comment:
can you document this method please? this is a good spot to explain what the different keeper threads are and what they do

Reply:
some inline comments explaining the logic would help too, e.g., watch_blocks returns a stream of new blocks that get processed by process_new_blocks . In this particular case, it's worth explaining *why* the code works like this. Why subscribe to new blocks vs the events directly? There must be some thinking behind that. Write down the reasoning in the code for future reference.

Reply:
actually generally think all of the functions in this file could use a nice doc comment on top. The current comments are either missing or aren't descriptive enough imo.